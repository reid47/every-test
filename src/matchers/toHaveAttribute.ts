import { formatValue } from './common';

export function toHaveAttribute(received: HTMLElement, attribute: string, value?: string) {
  if (arguments.length < 3) {
    const pass = !!received.getAttribute(attribute);
    const message = pass
      ? `Expected element not to have attribute "${attribute}".`
      : `Expected element to have attribute "${attribute}".`;

    return { pass, message };
  }

  const actualValue = received.getAttribute(attribute);
  const pass = actualValue === value;
  const message = pass
    ? `Expected element not to have attribute "${attribute}" with value ${formatValue(value)}.`
    : `Expected element to have attribute "${attribute}" with value ${formatValue(
        value
      )}, but actual value was ${formatValue(actualValue)}.`;

  return { pass, message };
}
