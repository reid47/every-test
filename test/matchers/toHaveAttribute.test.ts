import toHaveAttribute from '../../src/matchers/toHaveAttribute';

describe('toHaveAttribute', () => {
  test('passing', () => {
    const el = document.createElement('div');
    el.setAttribute('role', 'banner');
    el.setAttribute('data-hello', 'world');

    const result = toHaveAttribute(el, 'role', 'banner');
    expect(result.pass).toBeTruthy();
    expect(result.message).toMatchSnapshot();

    const result2 = toHaveAttribute(el, 'data-hello');
    expect(result2.pass).toBeTruthy();
    expect(result2.message).toMatchSnapshot();
  });

  test('failing', () => {
    const el = document.createElement('div');
    el.setAttribute('role', 'banner');
    el.setAttribute('data-hello', 'world');

    const result = toHaveAttribute(el, 'role', 'button');
    expect(result.pass).toBeFalsy();
    expect(result.message).toMatchSnapshot();

    const result2 = toHaveAttribute(el, 'data-something-else');
    expect(result2.pass).toBeFalsy();
    expect(result2.message).toMatchSnapshot();
  });
});
