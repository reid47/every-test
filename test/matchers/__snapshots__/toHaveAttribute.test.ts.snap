// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toHaveAttribute failing 1`] = `"Expected element to have attribute \\"role\\" with value \\"button\\", but actual value was \\"banner\\"."`;

exports[`toHaveAttribute failing 2`] = `"Expected element to have attribute \\"data-something-else\\"."`;

exports[`toHaveAttribute passing 1`] = `"Expected element not to have attribute \\"role\\" with value \\"banner\\"."`;

exports[`toHaveAttribute passing 2`] = `"Expected element not to have attribute \\"data-hello\\"."`;
